src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/InitializeStaticFieldsInline.cs

```diff
@@ -45,39 +46,53 @@ public abstract class InitializeStaticFieldsInlineAnalyzer<TLanguageKindEnum> :
 
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(CA1810Rule, CA2207Rule);
 
-        protected abstract bool InitialiesStaticField(SyntaxNode node, SemanticModel semanticModel, CancellationToken cancellationToken);
-        protected abstract TLanguageKindEnum AssignmentNodeKind { get; }
-
         public override void Initialize(AnalysisContext analysisContext)
         {
             analysisContext.EnableConcurrentExecution();
             analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
 
-            analysisContext.RegisterCodeBlockStartAction<TLanguageKindEnum>(codeBlockStartContext =>
+            analysisContext.RegisterOperationBlockStartAction(context =>
             {
-                if (!(codeBlockStartContext.OwningSymbol is IMethodSymbol methodSym) ||
-                    !methodSym.IsStatic ||
-                    methodSym.MethodKind != MethodKind.SharedConstructor)
+                if (!(context.OwningSymbol is IMethodSymbol method) ||
+                    !method.IsStatic ||
+                    method.MethodKind != MethodKind.StaticConstructor)
                 {
                     return;
                 }
 
                 var initializesStaticField = false;
-                codeBlockStartContext.RegisterSyntaxNodeAction(syntaxContext =>
+                var isStaticCtorMandatory = false;
+                context.RegisterOperationAction(context =>
                 {
-                    if (!initializesStaticField)
+                    if (isStaticCtorMandatory || initializesStaticField)
```

Comment:
This will lead to a race condition if first callback is for a regular static field assignment that sets initializesStaticField to true. We should remove this if statement.